File: samples/flush.exs
generate = fn ->
  refs = Enum.map(Enum.to_list(1..500), fn _ -> make_ref() end)
  pids = Enum.map(Enum.to_list(1..500), fn _ -> spawn(fn -> :ok end) end)

  Enum.zip(refs, pids)
  |> Enum.reduce([], fn {ref, pid}, acc -> [ref | [pid | acc]] end)
end

input_list = Enum.shuffle(generate.())
map_set = for i <- input_list, into: %{} do
  {i, nil}
end

Benchee.run(
  %{
    "1. flush_inbox/1" => fn {input, _} ->
      :flush.flush_inbox(input)
    end,
    "2. flush_inbox_sparse/1" => fn {input, _} ->
      :flush.flush_inbox_sparse(input)
    end,
    "3. flush_inbox_opt/1" => fn {input, _} ->
      :flush.flush_inbox_opt(input)
    end,
    "4. flush_with_set_remove/1" => fn {_, input} ->
      :flush.flush_with_set_remove(input)
    end,
    "5. flush_with_set_keep/1" => fn {_, input} ->
      :flush.flush_with_set_keep(input)
    end,
  },
  inputs: %{
    "0. 1k" => {input_list, map_set}
  },
  before_each: fn {input_list, _} = input ->
    for i <- Enum.shuffle(input_list) do
      cond do
        is_reference(i) ->
          send(self(), {i, nil, nil, nil})
        is_pid(i) ->
          send(self(), {:'EXIT', i, nil})
      end
    end

    :erlang.garbage_collect()

    input
  end,
  warmup: 5,
  time: 20,
  reduction_time: 2,
  memory_time: 2
)

File: src/flush.erl
-module(flush).

-export([flush_inbox/1, flush_inbox_sparse/1, flush_inbox_opt/1, flush_with_set_remove/1,
         flush_with_set_keep/1]).

flush_inbox([H | Rest]) ->
    receive
        {H, _, _, _} when is_reference(H) ->
            flush_inbox(Rest);
        {'EXIT', H, _} when is_pid(H) ->
            flush_inbox(Rest)
    after 0 ->
        flush_inbox(Rest)
    end;
flush_inbox([]) ->
    ok.

flush_inbox_sparse([Ref | Rest]) when is_reference(Ref) ->
    receive
        {Ref, _, _, _} ->
            flush_inbox_sparse(Rest)
    after 0 ->
        flush_inbox_sparse(Rest)
    end;
flush_inbox_sparse([Pid | Rest]) when is_pid(Pid) ->
    receive
        {'EXIT', Pid, _} ->
            flush_inbox_sparse(Rest)
    after 0 ->
        flush_inbox_sparse(Rest)
    end;
flush_inbox_sparse([]) ->
    ok.

flush_inbox_opt([H | Rest]) ->
    receive
        {Y, _, _, _} when is_reference(H), H =:= Y ->
            flush_inbox_opt(Rest);
        {'EXIT', Y, _} when is_pid(H), H =:= Y ->
            flush_inbox_opt(Rest)
    after 0 ->
        flush_inbox_opt(Rest)
    end;
flush_inbox_opt([]) ->
    ok.

flush_with_set_remove(Set) ->
    receive
        {Y, _, _, _} when is_reference(Y) andalso is_map_key(Y, Set) ->
            flush_with_set_remove(maps:remove(Y, Set));
        {'EXIT', Y, _} when is_pid(Y) andalso is_map_key(Y, Set) ->
            flush_with_set_remove(maps:remove(Y, Set))
    after 0 ->
        ok
    end.

flush_with_set_keep(Set) ->
    receive
        {Y, _, _, _} when is_reference(Y) andalso is_map_key(Y, Set) ->
            flush_with_set_keep(Set);
        {'EXIT', Y, _} when is_pid(Y) andalso is_map_key(Y, Set) ->
            flush_with_set_keep(Set)
    after 0 ->
        ok
    end.
